google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java

```diff
@@ -1757,6 +1762,103 @@ public ApiFuture<QuerySnapshot> get() {
     return get(null);
   }
 
+  ApiFuture<QueryProfileInfo<QuerySnapshot>> getQueryProfileInfo(QueryMode queryMode) {
+    final SettableApiFuture<QueryProfileInfo<QuerySnapshot>> result = SettableApiFuture.create();
+
+    RunQueryRequest.Builder request = RunQueryRequest.newBuilder();
+    request
+        .setStructuredQuery(buildQuery())
+        .setParent(options.getParentPath().toString())
+        .setMode(queryMode);
+
+    final List<QueryDocumentSnapshot> documentSnapshots = new ArrayList<>();
+
+    ResponseObserver<RunQueryResponse> observer =
+        new ResponseObserver<RunQueryResponse>() {
+          Timestamp readTime;
+          Struct planStruct = Struct.getDefaultInstance();
+          Struct statsStruct = Struct.getDefaultInstance();
+
+          @Override
+          public void onStart(StreamController streamController) {}
+
+          @Override
+          public void onResponse(RunQueryResponse response) {
+            if (response.hasDocument()) {
+              Document document = response.getDocument();
+              QueryDocumentSnapshot documentSnapshot =
+                  QueryDocumentSnapshot.fromDocument(
+                      rpcContext, Timestamp.fromProto(response.getReadTime()), document);
+              documentSnapshots.add(documentSnapshot);
+            }
+
+            if (readTime == null) {
+              readTime = Timestamp.fromProto(response.getReadTime());
+            }
+
+            if (response.hasStats()) {
+              if (response.getStats().hasQueryPlan()) {
+                planStruct = response.getStats().getQueryPlan().getPlanInfo();
+              }
+              if (response.getStats().hasQueryStats()) {
+                statsStruct = response.getStats().getQueryStats();
+              }
```

Comment:
nit: The same getter method is called multiple times. We should pull this out into a local variable.

```suggestion
              ResultSetStats stats = response.getStats();
              if (stats.hasQueryPlan()) {
                planStruct = stats.getQueryPlan().getPlanInfo();
              }
              if (stats.hasQueryStats()) {
                statsStruct = stats.getQueryStats();
              }
```